libraries/Microsoft.Bot.Builder.Azure/CosmosDbStorage.cs

```diff
@@ -0,0 +1,236 @@
+{
+    /// <summary>
+    /// Models IStorage around CosmosDB
+    /// </summary>
+    public class CosmosDbStorage : IStorage
+    {
+        private readonly string _databaseId;
+        private readonly string _collectionId;
+        private readonly Lazy<string> _collectionLink;
+        private readonly DocumentClient _client;
+
+        private static JsonSerializer _jsonSerializer = JsonSerializer.Create(new JsonSerializerSettings()
+        {
+            TypeNameHandling = TypeNameHandling.All
+        });
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CosmosDbStorage"/> class,
+        /// using the provided CosmosDB credentials, DatabaseId and CollectionId.
+        /// </summary>
+        /// <param name="serviceEndpoint">The endpoint Uri for the service endpoint from the Azure Cosmos DB service.</param>
+        /// <param name="authKey">The AuthKey used by the client from the Azure Cosmos DB service.</param>
+        /// <param name="databaseId">The Database ID.</param>
+        /// <param name="collectionId">The Collection ID.</param>
+        public CosmosDbStorage(Uri serviceEndpoint, string authKey, string databaseId, string collectionId, Action<ConnectionPolicy> connectionPolicyConfigurator = null)
+        {
+            if (serviceEndpoint == null)
+            {
+                throw new ArgumentNullException(nameof(serviceEndpoint), "Service EndPoint for CosmosDB is required.");
+            }
+
+            if (string.IsNullOrEmpty(authKey))
+            {
+                throw new ArgumentException("AuthKey for CosmosDB is required.", nameof(authKey));
+            }
+
+            if (string.IsNullOrEmpty(databaseId))
+            {
+                throw new ArgumentException("DatabaseId is required.", nameof(databaseId));
+            }
+
+            if (string.IsNullOrEmpty(collectionId))
+            {
+                throw new ArgumentException("CollectionId is required.", nameof(collectionId));
+            }
+
+            _databaseId = databaseId;
+            _collectionId = collectionId;
+
+            var version = GetType().Assembly.GetName().Version;
+            var connectionPolicy = new ConnectionPolicy()
+            {
+                UserAgentSuffix = $"Microsoft-BotFramework {version}"
+            };
+
+            connectionPolicyConfigurator?.Invoke(connectionPolicy);
+            _client = new DocumentClient(serviceEndpoint, authKey, connectionPolicy);
+
+            _collectionLink = new Lazy<string>(EnsureCollectionExist);
+        }
+
+        /// <summary>
+        /// Deletes the specified keys from storage.
+        /// </summary>
+        /// <param name="keys">List of keys to remove.</param>
+        /// <returns></returns>
+        public async Task Delete(params string[] keys)
+        {
+            // Ensure collection exists
+            var collectionLink = _collectionLink.Value;
+
+            var tasks = keys.Select(key =>
```

Comment:
Is there a way to do this w/o LINQ? Seems that would be the same amount of code, yet clearer and easier to follow.

Reply:
Updated for simplicity. Still, the idea is to map the keys into deletion Tasks, and wait for they completion:
````C#
            // Parallelize deletion
            var tasks = keys.Select(key =>
            {
                return _client.DeleteDocumentAsync(UriFactory.CreateDocumentUri(_databaseId, _collectionId, SanitizeKey(key)));
            });

            // await to deletion tasks to complete
            await Task.WhenAll(tasks).ConfigureAwait(false);
````